### PR TITLE
Add profile property API

### PR DESCRIPTION
**Pull request by kashike**
[**API**](https://github.com/SpongePowered/SpongeAPI/pull/974) | [Common](https://github.com/SpongePowered/SpongeCommon/pull/344) | [Forge](https://github.com/SpongePowered/SpongeForge/pull/455) | [Vanilla](https://github.com/SpongePowered/SpongeVanilla/pull/194)

This adds access to the property map that is associated with `GameProfile`s.

The property map is represented by a `Multimap` of `String`s to `ProfileProperty`s, although the keys of the multimap are ignored internally. Each `ProfileProperty` contains a name (non-null), value (non-null), and an optional (nullable) signature - if the signature is included, it **must** be a valid signature provided by Mojang which matches the value.
Textures (skins, capes, etc) are one item that is stored in the property map (used for resolving humanoid skins/capes/etc), although there may be other properties in the map, either from Mojang or other modifications/plugins.
Like entity AI, profile properties must be constructed through `GameProfileProperty`, either by extending the class or instantiating with your desired values. Attempting to add an extension of `ProfileProperty` will result in an exception being thrown at you.

Test plugin: https://gist.github.com/kashike/32e61d0ba0ecf76a5d04

### CODE REVIEW

File: src/main/java/org/spongepowered/api/profile/property/GameProfilePropertySuperclass.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.profile.property;
+
+import javax.annotation.Nullable;
+
+/**
+ * Internal use only.
+ */
+class GameProfilePropertySuperclass {
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/profile/property/GameProfilePropertySuperclass.java]: Implementation should not dictate how the API is designed IMO. Perhaps the property could be just an interface in the API and there be a factory to create instances

Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/profile/property/GameProfilePropertySuperclass.java]: This can just go in the implementation - no need for it in the API.

Then, you can use my superclass transformer to set the superclass at runtime.

File: src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package org.spongepowered.api.profile.property;
+
+import com.google.common.collect.Multimap;
+import org.spongepowered.api.profile.GameProfile;
+
+/**
+ * Represents a {@link String} to {@link ProfileProperty} map
+ * for {@link GameProfile}s.
+ */
+public interface ProfilePropertyMap extends Multimap<String, ProfileProperty> {
```
Comment (JBYoshi) [inline src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java]: Why not just use `Multimap<String, ProfileProperty>` (or, if needed, its immutable variant)?

**Comment by gabizou** [comment]
@kashike Can you provide a descriptive description about what the point of this API PR? What is it providing, what is the use case (in layman terms), etc.

**Comment by JBYoshi** [comment]
One possible usage could be setting skins on skulls.

File: src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java (hunk 2 above)
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java]: Why do you use a `String (key) -> { String (key), String (value) }` mapping anyway?
Why not a simple `String (key) -> String (value)` (multi) mapping.

`getProperty(String key) : String`
`getPropertySignature(String key) : Optional<String>`
(those emthods could also return a collection of value+signature tuples)
`getProperty(String key) : Collection<Tuple<StringValue,OptionalStringSignature>>`

**Comment by gabizou** [comment]
> One possible usage could be setting skins on skulls.

I get that, but again, this is why we have descriptions.

File: src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java (hunk 2 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java]: @JBYoshi `ProfilePropertyMap` is mixin'd to `PropertyMap`, used during serialisation/etc.
@ST-DDT This deals with `ProfileProperty` objects instead of the raw name/value/signature. In the implementation, the `PropertyMap` is a `Multimap<String, Property>`, we simply "extract" that to the API.

File: src/main/java/org/spongepowered/api/profile/property/GameProfileProperty.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+     * @param value The value of the property.
+     * @param signature The signature of the property.
+     */
+    public GameProfileProperty(String name, String value, @Nullable String signature) {
+        this.name = checkNotNull(name, "name");
+        this.value = checkNotNull(value, "value");
+        this.signature = signature;
+    }
+
+    /**
```
Comment (phroa) [inline src/main/java/org/spongepowered/api/profile/property/GameProfileProperty.java]: pretty sure you can leave these off

**Comment by kashike** [comment]
Could these property PRs be reviewed soon?

**Comment by kashike** [comment]
@gabizou I'm going to say this is done unless any comments come up.
I'm not entirely sure on making `GameProfile`'s `getUniqueId` and `getName` methods `Optional`, but as-is they can be null for mod-provided profiles.

File: src/main/java/org/spongepowered/api/profile/GameProfileCache.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.profile;
+
+import java.util.Collection;
+
+/**
+ * Represents a cache of {@link GameProfile}s.
+ */
+public interface GameProfileCache {
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: `GameProfileManager` already is a thing. There's [`getCachedProfiles()`](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/profile/GameProfileManager.java#L115)

File: src/main/java/org/spongepowered/api/profile/GameProfileManager.java [hunk 5]
```diff
@@ -99,20 +157,54 @@
...
+     * caused by a {@link ProfileNotFoundException} if the profile does not exist or
+     * an {@link IOException} if a network error occurred.</p>
+     *
+     * @param profile The profile to fill
+     * @param signed true if we should request that the profile data be signed
      * @param useCache true to perform a cache lookup first
      * @return The result of the request
      */
-    ListenableFuture<Collection<GameProfile>> getAllById(Iterable<UUID> uniqueIds, boolean useCache);
+    ListenableFuture<GameProfile> fill(GameProfile profile, boolean signed, boolean useCache);
```
Comment (simon816) [inline src/main/java/org/spongepowered/api/profile/GameProfileManager.java]: I would argue that the concept of 'filling' a game profile is an implementation detail and should not be needed in the API.

Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/profile/GameProfileManager.java]: I don't really think so. It's perfectly possible for a plugin to construct a `GameProfile` on its own, with just a UUID (or UUID and name). There should be a way to fill it with the appropriate profile properties.

Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfileManager.java]: As @Aaron1011 said, this is required for newly constructed profiles.

File: src/main/java/org/spongepowered/api/profile/GameProfileCache.java (hunk 4 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: To prevent breaks in the future if we want to add methods to control the cache - removing, clearing, getting size, etc.

Comment (gabizou) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: There wouldn't be breaks in the future because it's a manager and not a service. Plugins aren't going to be able to replace the `GameProfileManager`, so, it's just going to get new methods, but no methods are going to be added for plugins to implement. Yes, it's a partial breaking change, but it's only really additions for the context of where plugins are concerned.

Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: breaks when moving methods to a new cache interface as I did. filling up a manager with cache related methods will get ugly.

Comment (gabizou) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: But again, if you simply keep the method to the `GameProfileManager` there is no breakage. It's silly to add a new interface to interface with a direct cache object that is being only partially exposed.

Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: It is silly to prevent a future break when this is breaking already?

Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfileCache.java]: whatever, reverted

File: src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java (hunk 2 above)
Comment (simon816) [inline src/main/java/org/spongepowered/api/profile/property/ProfilePropertyMap.java]: Use of a mixin is an implementation detail and should have no effect on the API. Anyway, you can just return the `PropertyMap` since it implements `Multimap` anyway

**Comment by simon816** [comment]
How about adding `boolean isFilled()` to `GameProfile` so we know if it's got it's extended properties filled or not.

File: src/main/java/org/spongepowered/api/profile/GameProfile.java [hunk 6]
```diff
@@ -37,9 +42,28 @@
...
 
     /**
-     * Gets the player's last known username.
+     * Creates a {@link GameProfile} from the provided ID and name.
+     *
+     * @param uniqueId The unique ID
+     * @param name The name
+     * @return The created profile
+     */
+    static GameProfile of(UUID uniqueId, String name) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: If the name is optional, should we consider that the name should be `@Nullable`? @SpongePowered/developers thoughts?

File: src/main/java/org/spongepowered/api/profile/property/ProfileProperty.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+/**
+ * Represents a property in a {@link GameProfile}'s {@link ProfilePropertyMap}.
+ *
+ * <p>This interface should not be implemented. If you desire to create your own, see
+ * {@link #of(String, String)} or {@link #of(String, String, String)}.</p>
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/profile/property/ProfileProperty.java]: Use `@see #of(String, String)` annotation instead please.

File: src/main/java/org/spongepowered/api/profile/GameProfile.java (hunk 6 above)
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: I would say yes, as it's what plugins devs would expect after seeing the `Optional` name. There's also a usecase: A plugin might create a `GameProfile`, and then fill it to get any data they didn't know (such as the name).

Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: done

File: src/main/java/org/spongepowered/api/profile/property/ProfileProperty.java (hunk 7 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/property/ProfileProperty.java]: done

**Comment by kashike** [comment]
@simon816 added
@gabizou done

File: src/main/java/org/spongepowered/api/profile/GameProfile.java [hunk 8]
```diff
@@ -37,9 +44,40 @@
...
+     *
+     * @return The property map
+     */
+    ProfilePropertyMap getPropertyMap();
+
+    /**
+     * Checks if this profile is filled.
      *
-     * @return The player's last known username
+     * <p>A filled profile has no empty fields.</p>
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: Can you clarify what this means with regard to `ProfileProperties`?

File: src/main/java/org/spongepowered/api/profile/GameProfile.java [hunk 9]
```diff
@@ -37,9 +44,40 @@
...
+     * @return The associated name if present, otherwise {@link Optional#empty()}
+     */
+    Optional<String> getName();
+
+    /**
+     * Gets the property map for this profile.
+     *
+     * @return The property map
+     */
+    ProfilePropertyMap getPropertyMap();
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: You should clarify that this is a live map (unless you're going to add a setter method).

File: src/main/java/org/spongepowered/api/profile/GameProfile.java (hunk 8 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: done

File: src/main/java/org/spongepowered/api/profile/GameProfile.java (hunk 9 above)
Comment (kashike) [inline src/main/java/org/spongepowered/api/profile/GameProfile.java]: done

**Comment by kashike** [comment]
https://github.com/SpongePowered/SpongeAPI/commit/e022fc0f05df998ed5c0e371f8d65e916872c560
